### ISSUE TITLE
Running dependents (eg always run task B after task A)

### ISSUE THREAD

**Issue by tremby** [gulpjs/gulp#1406]
Gulp 3 gives control over dependencies which need to run before a given task. With those, I can say task B depends on task A. If I then run A, A runs and then we exit. If I run B, A runs and then B runs, and then we exit. But if A has already run and its output is valid, during development when some files B operates on change I may want to just run B without running A. So I'd rather describe the tasks in terms of dependents rather than dependencies. In other words, I'd like to tell Gulp that whenever A runs, B should run afterwards.

If the chain is longer, like A through E, and files are being edited such that only E strictly needs to be run, the entire chain will still run since E depends on D, D depends on C, C depends on B, B depends on A. Say I edit a file such that C needs to run, I do need D and E to run afterwards, but not A and B.

Is there an elegant way to set this up in Gulp 3? At the moment I have a bunch of `run-sequence`s, and it's somewhat unpleasant to maintain, and in certain cases inefficient since I can't so easily dictate which sets of tasks can run in parallel, at least not with a fine level of control.

Or is this something Gulp 4 addresses at all?

I believe even with `npm run` (package.json scripts) you can define tasks which need to be run after other tasks. I feel that I must be missing something here.

**Comment by sprynm** [comment]
Parallel vs series in Gulp 4 may be suitable.

		gulp.task('A', function() {
		  ...
		}
		gulp.task('B', function() {
		  ...
		}

		// runs A then B
		 gulp.task('buildAthenB', gulp.series( A, B ) );

		// runs A and B concurrently
		 gulp.task('buildAthenB', gulp.parallel( A, B ) );

From there you could set a watch on the output of A to trigger B.

You can even mix and match to fit your dependency needs:

        gulp.task('build', gulp.series(
            'A',                         //1
            gulp.parallel(               //2
                'B', 
                gulp.series(
                    'D',
                    'F'
                )
            ),
            gulp.parallel(               //3 
                'C', 
                'E'
            ),
            function() {                 //4
                notifier.notify({ message: "All files are built" }) 
            }
        ));

`gulp build` now runs: 
1. A
2. B & D then F
3. C & E
4. Notice

**Comment by phated** [comment]
Well written @sprynm

**Comment by tremby** [comment]
I get that and that's nice and I can certainly do what I need to do with it, but I think it'll result in a lot of code and a lot of tasks. When a new task is added I need toI think there will be a lot of redundancy in the gulpfile.

I think it'd be more elegant and much more maintainable to be able to

- Define pre- and post- (and during-? haven't thought that through) dependencies along with any given task
- Automatically run dependencies when triggering a task
- Make it possible to force a task to run *without* its dependencies running
- Define a gulp task which runs a given set of tasks, but automatically determines their interdependencies and so decides the best order to run them in, with as much parallellization as possible
- Obviously still allow the gulp.parallel and gulp.series for fine control where necessary
